### PR TITLE
gadget/update.go: add buildNewVolumeToDeviceMapping for existing devices

### PR DESCRIPTION
**Pull request by anonymouse64**
For devices that don't have a disk-mapping.json, we need this method to
construct the traits for the system-boot containing volume only to allow
gadget asset updates to be performed on specifically that volume only.

The naive check to see if the device has encryption or not may be too 
naive, not sure...

### CODE REVIEW

File: gadget/update.go [hunk 1]
```diff
@@ -665,6 +667,143 @@ func DiskTraitsFromDeviceAndValidate(expLayout *LaidOutVolume, dev string, opts
...
+		}
+	}
+	if systemBootVolume == "" {
+		// didn't find system-boot anywhere somehow
+		if preUC20 {
+			logger.Noticef("WARNING: could not identify disk for gadget asset update of volume %s: unable to find any volume with system-boot role on it", systemBootVolume)
+			return nil, errSkipUpdateProceedRefresh
+		}
+		// shouldn't be possible on UC20
+		return nil, fmt.Errorf("unable to find any volume with system-boot, gadget is broken")
```
Comment (pedronis) [inline gadget/update.go:702]: usual comment to try to start errors with "cannot" as much as possible

File: gadget/update.go [hunk 2]
```diff
@@ -665,6 +667,143 @@ func DiskTraitsFromDeviceAndValidate(expLayout *LaidOutVolume, dev string, opts
...
+
+	// setup encrypted structure information to perform validation if this
+	// device used encryption
+	if !preUC20 {
+		// TODO: this needs to check if the specified partitions are ICE when
+		// we support ICE too
+
+		// TODO: is this a good enough check to know if encryption was turned
+		// on? note that we can't import the boot pkg here
+		keysPattern := filepath.Join(dirs.SnapFDEDir, "*.key")
```
Comment (pedronis) [inline gadget/update.go:764]: I would use marker perhaps. Most of the .key will go away once we make recovery key optional, so at most we would have ubuntu-save.key

**Review by pedronis — COMMENTED** [review]
couple of initial comments

File: gadget/update.go (hunk 1 above)
Comment (anonymouse64) [reply gadget/update.go:702]: yes, I don't know why my brain refuses to remember this every time I go to write an error message

File: gadget/update.go [hunk 3]
```diff
@@ -665,6 +667,143 @@ func DiskTraitsFromDeviceAndValidate(expLayout *LaidOutVolume, dev string, opts
 	}, nil
 }
 
+// unable to proceed with the gadget asset update, but not fatal to the refresh
+// operation itself
+var errSkipUpdateProceedRefresh = errors.New("could not identify disk for gadget asset update")
+
+// buildNewVolumeToDeviceMapping builds a DiskVolumeDeviceTraits for only the
+// system-boot role containing volume when we cannot load an existing traits
```
Comment (bboozzoo) [inline gadget/update.go:675]: found it a bit convoluted, how about:
```suggestion
// buildNewVolumeToDeviceMapping builds a DiskVolumeDeviceTraits only for the
// volume containing the system-boot role, when we cannot load an existing traits
```

File: gadget/update.go [hunk 4]
```diff
@@ -665,6 +667,143 @@ func DiskTraitsFromDeviceAndValidate(expLayout *LaidOutVolume, dev string, opts
...
+			// at least should have ubuntu-data encrypted
+			validateOpts.ExpectedStructureEncryption = map[string]StructureEncryptionParameters{
+				"ubuntu-data": {Method: EncryptionLUKS},
+			}
+
+			// check if ubuntu-save exists to know whether to also assume
+			// encryption for ubuntu-save too - this should always happen since
+			// we require ubuntu-save for encryption, but there was a time
+			// during UC20 development where you could have encryption without
+			// ubuntu-save, so there is a small chance such devices could have
```
Comment (bboozzoo) [inline gadget/update.go:781]: I think such gadget would fail validation now, so we would not reach here anyway.

Comment (anonymouse64) [reply gadget/update.go:781]: do you think I should get rid of this entirely and just always assume that if encryption is turned on for a device that it has ubuntu-save ?

File: gadget/update.go (hunk 2 above)
Comment (anonymouse64) [reply gadget/update.go]: changed to use the marker file instead

File: gadget/update.go (hunk 1 above)
Comment (anonymouse64) [reply gadget/update.go]: fixed

File: gadget/update.go [hunk 5]
```diff
@@ -665,6 +668,139 @@ func DiskTraitsFromDeviceAndValidate(expLayout *LaidOutVolume, dev string, opts
...
+
+	if dev == "" {
+		// couldn't find a disk at all, pre-UC20 we just warn about this
+		// but let the update continue
+		if preUC20 {
+			logger.Noticef("WARNING: could not identify disk for gadget asset update of volume %s", systemBootVolume)
+			return nil, errSkipUpdateProceedRefresh
+		}
+		// fatal error on UC20+
+		return nil, fmt.Errorf("could not identify disk for gadget asset update of volume %s", systemBootVolume)
```
Comment (pedronis) [inline gadget/update.go:748]: same about cannot, sorry I should made punctual comments, instead of a general one

File: gadget/update.go (hunk 4 above)
Comment (pedronis) [reply gadget/update.go:778]: if we are worried about device before we shipped UC 20, I think it's ok to ignore that case, I'm sure we have other bits of code that simply assume ubuntu-save if encrpytion is on

**Review by pedronis — COMMENTED** [review]
re-reviewed

File: gadget/update.go (hunk 4 above)
Comment (anonymouse64) [reply gadget/update.go:778]: ok, I will simplify this check here

File: gadget/update.go (hunk 5 above)
Comment (anonymouse64) [reply gadget/update.go]: hopefully got them all this time

**Comment by anonymouse64** [comment]
I accidentally refactored out an important part of this code, I need to re-add it, marking blocked until I do so (hopefully should be quick)

**Comment by anonymouse64** [comment]
ok, this PR is ready again I pushed back the important bit of code that handled old gadgets, see the commit message for full details

**Review by bboozzoo — APPROVED** [review]
LGTM

**Review by pedronis — APPROVED** [review]
thank you

**Pull request merged**